src/NUnitFramework/tests/Internal/ExceptionHelperOutputExceptionDataTests.cs

```diff
@@ -0,0 +1,161 @@
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+// ***********************************************************************
+// Copyright (c) 2019 Charlie Poole
+// 
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+
+namespace NUnit.Framework.Internal 
+{
+    public static class ExceptionHelperOutputExceptionDataTests
+    {
+        [Test]
+        public static void AppendsDataItemsToExceptionMessage()
+        {
+            var exception = new Exception("blah");
+            exception.Data["data-prop"] = "data-value";
+            
+            var message = ExceptionHelper.BuildMessage(exception);
+            Assert.That(message, Contains.Substring("data-prop"));
+            Assert.That(message, Contains.Substring("data-value"));
+        }
+
+        [Test]
+        public static void CanDisable()
+        {
+            var exception = new Exception("blah");
+            exception.Data["data-prop"] = "data-value";
+
+            string message;
+            try
+            {
+                ExceptionHelper.OutputExceptionDataProperty = false;
+                message = ExceptionHelper.BuildMessage(exception);
+            }
+            finally
+            {
+                ExceptionHelper.OutputExceptionDataProperty = true;
+            }
+            
+            Assert.That(message, !Contains.Substring("data-prop"));
+            Assert.That(message, !Contains.Substring("data-value"));
+        }
```

Comment:
I don't think we need to test turning it on and off for the same exception. The `finally` block is still needed because we changed a static property but as noted elsewhere I don't think we want to use a static.

I'd also like to see a test with a few values and have one test that checks the actual format of the message. The latter is more fragile, of course, but we should make sure the message is exactly as expected in one case anyway.

For completeness, let's have a test with nothing in the data dictionary. Of course that's tested by practically every other test we have, but it would be good to make it explicit here.

Reply:
if we make the feature not optional, this test can go away.

Reply:
Is the message format a contract?  Are there tests for the format I can follow?  I didn't see any.  Typically, I just verify message parts to reduce fragility.

Reply:
I can add an explicit test for the empty dictionary.  As you said, it's covered by all the other tests so I didn't think it necessary.  Explicit is good.